dftimewolf/lib/collectors/gsheets.py

```diff
@@ -0,0 +1,258 @@
+      if not credentials or not credentials.valid:
+        if credentials and credentials.expired and credentials.refresh_token:
+          credentials.refresh(Request())
+        else:
+          secrets_path = os.path.join(
+              os.path.expanduser('~'), self._CLIENT_SECRET_FILENAME)
+          if not os.path.exists(secrets_path):
+            error_message = (
+                'No OAuth application credentials available to access google '
+                'sheets. Please generate OAuth application credentials (see '
+                'https://developers.google.com/sheets/api/guides/authorizing) '
+                'and save them to {0:s}.').format(secrets_path)
+            self.ModuleError(error_message, critical=True)
+          flow = InstalledAppFlow.from_client_secrets_file(
+              secrets_path, self.SCOPES)
+          credentials = flow.run_console()
+
+        # Save the credentials for the next run
+        with open(credentials_path, 'w') as token_file:
+          token_file.write(credentials.to_json())
+
+    return credentials
+
+  def _BuildSheetsResource(self,
+                           credentials: Credentials) -> discovery.Resource:
+    """Builds a Google Sheets resource object to use to request logs.
+
+    Args:
+      credentials: Google API credentials
+
+    Returns:
+      A resouce object for interacting with the Google Sheets API.
+    """
+    return discovery.build('sheets', 'v4', credentials=credentials)
+
+  def _ValidateSpreadSheetId(self, spreadsheet: str) -> str:
+    """Extract and validate spreadsheet ID.
+
+    Args:
+      spreadsheet: ID or URL of the sheetspread,
+
+    Returns:
+      spreadsheet ID
+    """
+    spreadsheet_match = re.search(r'.*?([01][0-9A-Za-z_-]{20,}).*', spreadsheet)
+    if not spreadsheet_match:
+      self.ModuleError(
+          'spreadsheet id is not in the correct format {0:s}.'.format(
+              spreadsheet),
+          critical=True)
+      return ""
+
+    return spreadsheet_match.group(1)
+
+  def _ExtractEntiresFromSheet(self,
+                               spreadsheet_id: str,
+                               sheet_title: str) -> DataFrame:
+    """Extract entries from the sheet inside the spreadsheet.
+
+    Args:
+      spreadsheet_id: ID of the spreadsheet to pull data from
+      sheet_title: Title of the sheet inside the spreadsheet to parse.
+
+    Returns:
+        Dataframe with entries from sheet inside the spreadsheet
+    """
+
+    resource = self._BuildSheetsResource(self._credentials)
+    # Pylint can't see the spreadsheets method.
+    # pylint: disable=no-member
+    sheet_content_result = resource.spreadsheets().values().get(
+        spreadsheetId=spreadsheet_id, range=sheet_title).execute()
+    values = sheet_content_result.get('values', [])
+
+    if not values:
+      self.logger.warning('No data found in sheet "{0:s}".'.format(sheet_title))
+      return None
```

Comment:
The type hinting says this function returns a DataFrames, so you should be returning an empty dataframe here, or add None as possible return type.

Reply:
Done